### PR TITLE
feat(store-api): Add cart to store-api

### PR DESCRIPTION
**Pull request by tlgimenes**
## What's the purpose of this pull request?
This PR adds an optimistic cart implementation to store-api.

## How it works? 
The optimistic cart we have works via state validation. This means that, we build a state of the cart on the browser, and then, the backend validates this state. If this state is valid, the backend signalizes this state is ok and nothing changes on the frontend. If, however, this state is not valid, a.k.a, this product is unavailable, a promotion was applied, etc, the backend sends the new cart state to the frontend. An example of this process can be seen below: 

image

To implement the cart, I used the schema.org's type `Order`. Also, another thing this PR brings is `input` fields. We didn't really use them before, but now I needed a standard for writing them. I added an `I` prefix to them, so writing an `input` for a `type` can be achieved by:
`type StoreProduct` -> `type IStoreProduct`.

Also, another thing this PR brings is the usage of [dataloaders](https://github.com/graphql/dataloader). This was necessary for loading skus and simulations without overwhelming the search/checkout API.

## How to test it?
To test it, open our PR on base.store and check that all promotions are applied correctly. Also, if you want you can change the base account name to `carrefourbrfood` and check all edge cases in there. 

https://github.com/vtex-sites/base.store/pull/36

### CODE REVIEW

File: packages/store-api/src/platforms/vtex/clients/commerce/types/OrderForm.ts [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  listPrice: number
+  manualPrice: null
+  manualPriceAppliedBy: null
+  sellingPrice: number
+  rewardValue: number
+  isGift: boolean
+  additionalInfo: AdditionalInfo
+  preSaleDate: null
+  productCategoryIds: string
+  productCategories: { [key: string]: string }
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/clients/commerce/types/OrderForm.ts:96]: Record<string, string>, what do you think?

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+} from '../clients/commerce/types/OrderForm'
+import type { Context } from '..'
+
+const getId = (item: IStoreOffer) =>
+  [item.itemOffered.sku, item.seller.identifier, item.price].join('::')
+
+const orderFormItemToOffer = (
+  item: OrderFormItem,
+  index?: number
+): IStoreOffer & { index?: number } => ({
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:15]: What do you think about creating an interface (or even type) to reuse this code?
```ts
type IndexedType  = { index?: number }
```
```suggestion
): IStoreOffer & IndexedType => ({
```
Name things isn't one of my best skills, as you can see.

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  itemOffered: {
+    sku: item.id,
+    image: [],
+    name: item.name,
+  },
+  index,
+})
+
+const offerToOrderItemInput = (
+  offer: IStoreOffer & { index?: number }
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:29]: Add IndexedType here too.

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        uniqueId,
+        quantity,
+        seller,
+        sellingPrice,
+        availability,
+      })
+    ),
+  })
+
+  return JSON.stringify(pick(of1)) === JSON.stringify(pick(of2))
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:62]: Do you think comparing a big string is a good idea, isn't it?

I have a thin algorithm that is a little bit performant:

```ts
const comparison = (object1: any, object2: any) => {
  const keys = Object.keys(object1);

  for (let i = 0; i < keys.length; i++) {
    if (
      typeof object1[keys[i]] === "string" ||
      typeof object1[keys[i]] === "number"
    ) {
      if (object1[keys[i]] !== object2[keys[i]]) {
        return false;
      }
    }

    if (
      (Array.isArray(object1[keys[i]]) ||
        typeof object1[keys[i]] === "object") &&
      !comparison(object1[keys[i]], object2[keys[i]])
    ) {
      return false;
    }
  }

  return true;
};
```

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+ * This resolver implements the optimistic cart behavior. The main idea in here
+ * is that we receive a cart from the UI (as query params) and we validate it with
+ * the commerce platform. If the cart is valid, we return null, if the cart is
+ * invalid according to the commerce platform, we return the new cart the UI should use
+ * instead
+ *
+ * The algoritm is something like:
+ * 1. Fetch orderForm from VTEX
+ * 2. Compute delta changes between the orderForm and the UI's cart
+ * 3. Update the orderForm accordingly
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:75]: Why not just return the new the orderform on this step?

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+const getId = (item: IStoreOffer) =>
+  [item.itemOffered.sku, item.seller.identifier, item.price].join('::')
+
+const orderFormItemToOffer = (
+  item: OrderFormItem,
+  index?: number
+): IStoreOffer & { index?: number } => ({
+  listPrice: item.listPrice / 100,
+  price: item.sellingPrice / 100,
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:17]: I prefer to transfer this data over the net with a natural value, not a float.

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts (hunk 4 above)
Comment (igorbrasileiro) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:62]: https://codesandbox.io/s/typescript-forked-o67fh?file=/src/index.ts:0-537

File: packages/store-api/src/platforms/vtex/loaders/simulation.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  Simulation,
+} from '../clients/commerce/types/Simulation'
+import type { Options } from '..'
+import type { Clients } from '../clients'
+
+export const getSimulationLoader = (_: Options, clients: Clients) =>
+  new DataLoader<PayloadItem[], Simulation>(
+    async (items) => {
+      const simulated = await clients.commerce.checkout.simulation({
+        items: [...items.flat()],
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/loaders/simulation.ts:14]: This request to the simulation API doesn't respect the regionalization feature. You should add 
```ts
   items: [...items.flat()],
   shippingData: {
     logisiticInfo: [{ itemIndex: index of item on items list, regionId: "id" }],
   }

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    ),
+  })
+
+  return JSON.stringify(pick(of1)) === JSON.stringify(pick(of2))
+}
+
+/**
+ * This resolver implements the optimistic cart behavior. The main idea in here
+ * is that we receive a cart from the UI (as query params) and we validate it with
+ * the commerce platform. If the cart is valid, we return null, if the cart is
```
Comment (victorhmp) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:68]: Not so sure about returning `null` to signify that "no changes were made", or "the cart is valid". 

Why not add a field to the returned `orderForm` like `updated`, which would be a boolean value? Then we always return back the `orderForm`, but with an easy way for the frontend to check if the cart should be updated or not

File: packages/store-api/src/platforms/vtex/loaders/sku.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+        const index = indexById[sku.id]
+
+        sorted[index] = enhanceSku(sku, product)
+      }
+
+      return sorted
+    },
+    {
+      maxBatchSize: 99, // Max allowed batch size of Search API
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/loaders/sku.ts:46]: If this maxBatchSize shouldn't be changed, put a warning `Warning: Don't change this value, this the max allowed batch size of Search API`

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+  // Step4: Apply delta changes to order form
+  const updatedOrderForm = await commerce.checkout.updateOrderFormItems({
+    id: orderForm.orderFormId,
+    orderItems: changes,
+  })
+
+  // Step5: If no changes detected before/after updating orderForm, the order is validated
+  if (equals(orderForm, updatedOrderForm)) {
+    return null
```
Comment (victorhmp) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:145]: With the change I proposed a few lines above, this condition would just return the same orderForm to the client, but with the `updated` property set to `false`.

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  // Step4: Apply delta changes to order form
+  const updatedOrderForm = await commerce.checkout.updateOrderFormItems({
+    id: orderForm.orderFormId,
+    orderItems: changes,
+  })
+
+  // Step5: If no changes detected before/after updating orderForm, the order is validated
+  if (equals(orderForm, updatedOrderForm)) {
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:144]: I didn't get it the advantage of return null here.

File: packages/store-api/src/typeDefs/index.ts [hunk 12]
```diff
@@ -34,6 +38,10 @@ export const typeDefs = [
   ProductGroup,
   Organization,
   AggregateOffer,
+  Mutation,
```
Comment (victorhmp) [inline packages/store-api/src/typeDefs/index.ts:41]: This is just a nitpick, but how about moving this `Mutation` type to below `Query`? 😂

**Review by victorhmp — CHANGES_REQUESTED** [review]

**Review by igorbrasileiro — CHANGES_REQUESTED** [review]
Good job overall. This PR is a little bit huge, so I might not see problems by mistake.
I left some comments

File: packages/store-api/src/platforms/vtex/loaders/simulation.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+      })
+
+      const itemsIndices = items.reduce(
+        (acc, curr) => [...acc, curr.length + acc[acc.length - 1]],
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/loaders/simulation.ts:18]: Seeing the PayloadItem interface, it doesn't have any property called length. Can you explain why you access curr.length?

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts (hunk 4 above)
Comment (tlgimenes) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:64]: I can try to use [fast-deep-equal](https://www.npmjs.com/package/fast-deep-equal), what do you think?

File: packages/store-api/src/platforms/vtex/clients/commerce/index.ts [hunk 14]
```diff
@@ -28,18 +37,56 @@ export const VtexCommerce = (opts: Options) => {
     checkout: {
       simulation: (
         args: SimulationArgs,
-        options: SimulationOptions = { sc: '1' }
+        simulationOptions: SimulationOptions = { sc: channel }
```
Comment (emersonlaurentino) [inline packages/store-api/src/platforms/vtex/clients/commerce/index.ts:40]: `sc` is `salesChannel`? if yes, can we use the full name instead of the initials?

File: packages/store-api/src/platforms/vtex/clients/commerce/index.ts [hunk 15]
```diff
@@ -28,18 +37,56 @@ export const VtexCommerce = (opts: Options) => {
...
-            headers: {
-              'content-type': 'application/json',
-            },
+          }
+        )
+      },
+      orderForm: ({
+        id,
+        refreshOutdatedData = true,
+        sc = channel,
```
Comment (emersonlaurentino) [inline packages/store-api/src/platforms/vtex/clients/commerce/index.ts:55]: the same as my previous comment

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts (hunk 4 above)
Comment (emersonlaurentino) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:64]: Yes, is better to use this lib.

File: packages/store-api/src/platforms/vtex/loaders/simulation.ts (hunk 7 above)
Comment (igorbrasileiro) [reply packages/store-api/src/platforms/vtex/loaders/simulation.ts]: Of course, no problem.

File: packages/store-api/src/platforms/vtex/clients/commerce/index.ts (hunk 14 above)
Comment (igorbrasileiro) [reply packages/store-api/src/platforms/vtex/clients/commerce/index.ts:40]: I agree with

File: packages/store-api/src/platforms/vtex/clients/commerce/index.ts [hunk 16]
```diff
@@ -28,18 +37,56 @@ export const VtexCommerce = (opts: Options) => {
...
+      updateOrderFormItems: ({
+        id,
+        orderItems,
+        allowOutdatedData = 'paymentData',
+        sc = channel,
+      }: {
+        id: string
+        orderItems: OrderFormInputItem[]
+        allowOutdatedData?: 'paymentData'
+        sc?: string
```
Comment (igorbrasileiro) [inline packages/store-api/src/platforms/vtex/clients/commerce/index.ts:80]: Maybe here too, the same comment @emersonlaurentino did.

**Review by igorbrasileiro — APPROVED** [review]

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts (hunk 8 above)
Comment (tlgimenes) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:72]: Just to save data. These carts can be big and it would be nice to save a little bit of bandwidth when the cart is valid

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts (hunk 5 above)
Comment (tlgimenes) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:79]: We need to update the orderr form on the VTEX API

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts (hunk 8 above)
Comment (tlgimenes) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:72]: What do you think?

Comment (victorhmp) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:72]: Yeaah, makes sense! 

I don't know how users will react to this hahahaha they might not find it weird. But regardless, from a performance standpoint, I agree. This is a great way to save data when dealing with big carts 😄  Especially taking into account poor mobile connections.

File: packages/store-api/src/platforms/vtex/resolvers/validateCart.ts (hunk 11 above)
Comment (tlgimenes) [reply packages/store-api/src/platforms/vtex/resolvers/validateCart.ts:148]: discussion at: https://github.com/vtex/faststore/pull/963#discussion_r720489627

**Review by victorhmp — APPROVED** [review]
